File: app/src/test/java/com/pomo/sync/timer/ActivePhaseTimerTest.kt
package com.pomo.sync.timer

import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

public class ActivePhaseTimerTest {
    @Test
    public fun startPauseResumeDeriveRemainingFromProjection() {
        var now = 1_000L
        val timer = ActivePhaseTimer("device-a") { now }
        val plan = PhasePlan(PhaseKind.WORK, 60_000, null)
        timer.start(plan, "phase-1")
        assertEquals(60_000, timer.remainingMillis())
        now = 1_000 + 10_000
        assertEquals(50_000, timer.remainingMillis())
        timer.pause()
        now = 1_000 + 30_000
        assertEquals(50_000, timer.remainingMillis())
        timer.resume()
        now = 1_000 + 35_000
        assertEquals(45_000, timer.remainingMillis())
        val completed = timer.complete()
        assertTrue(completed.completedOperationIds.isNotEmpty())
        assertEquals(0, timer.remainingMillis())
    }
}

File: extension/tests/sync/activePhaseTimer.test.ts
import { describe, expect, test } from "bun:test";
import { ActivePhaseTimer } from "../../src/sync/timer/activePhaseTimer";

describe("active phase timer", () => {
  test("start pause resume derive remaining from projection", () => {
    let now = 1_000;
    const timer = new ActivePhaseTimer("device-a", () => now);
    timer.start({ kind: "WORK", durationMillis: 60_000, tagId: null }, "phase-1");
    expect(timer.remainingMillis()).toBe(60_000);
    now = 11_000;
    expect(timer.remainingMillis()).toBe(50_000);
    timer.pause();
    now = 31_000;
    expect(timer.remainingMillis()).toBe(50_000);
    timer.resume();
    now = 36_000;
    expect(timer.remainingMillis()).toBe(45_000);
    const completed = timer.complete();
    expect(completed.completedOperationIds.size).toBeGreaterThan(0);
    expect(timer.remainingMillis()).toBe(0);
  });
});
